### PR TITLE
Feature: read receipts

### PR DESCRIPTION
**Pull request by jpjpjp**
# Pull Request Template

## Description

This PR adds read receipt functionality to the SDK, specifically it adds 3 new functions:
* memberships.updateLastSeen()  -- sends an acknowledgement
* memberships.listWithReadStatus() -- adds read info to the membership list for a space
* rooms.listWithReadStatus -- return a list of spaces for all users with lastActivityDate and lastSeenDate

**TO DO BEFORE MERGING**
* node tests aren't working for rooms and memberships.  Complains that conversations is an unknown service.  
* remove messages.markAsReadFunction()

Fixes # (issue)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Test Coverage

Test coverage for the new functions have been added plugin-memberships and plugin-rooms tests.  WIth that said they are failing with an "Error: conversation is not a known service".   Any help on resolving this problem from reviewers woudl be appreciated!

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -20,7 +20,8 @@
         ],
         "node": "6.0"
       }
-    }]
+    }],
+    "stage-0"
```
Comment (andrewholsted) [inline .babelrc]: Why did you need to bring this in? Not necessarily against it, but I think it has larger implications for the sdk build steps

File: packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js [hunk 2]
```diff
@@ -269,6 +269,10 @@ const Conversation = SparkPlugin.extend({
           }, omit(options, 'id', 'user', 'url'))
         };
 
+        if ('participantAckFilter' in options) {
+          delete params.qs.includeParticipants;
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js]: typically I delete the things I check for in the `if` block. What's the logic here? A code comment would be great.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 3]
```diff
@@ -2,13 +2,15 @@
  * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
  */
 
+import '@webex/internal-plugin-conversation';
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: It looks like this is also included in /plugin-memberships/src/index.js. We shouldn't need it both places.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 4]
```diff
@@ -321,6 +323,89 @@ const Memberships = SparkPlugin.extend({
...
+   *
+   * This function may be deprecated when this info is provided in the membership
+   * objects returned in the list function.
+   *
+   *
+   * @instance
+   * @memberof Memberships
+   * @param {Object} options
+   * @param {string} options.roomId
+   * @returns {Promise<MembershipObjectList>}
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Epic js doc! 👍

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 5]
```diff
@@ -321,6 +323,89 @@ const Memberships = SparkPlugin.extend({
...
+   */
+  listWithReadStatus(options) {
+    const conversation = {
+      id: deconstructHydraId(options.roomId).id
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: This is always assuming a room id is base64 encoded, yeah? On web, we don't have the public hydra ids. Not sure if we need a check for that or not?

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 6]
```diff
@@ -321,6 +323,89 @@ const Memberships = SparkPlugin.extend({
...
+              personOrgId: constructHydraId(hydraTypes.ORGANIZATION,
+                participant.orgId),
+              isModerator: participant.roomProperties.isModerator,
+              isMonitor: false,
+              roomType: resp.tags.includes(SDK_EVENT.INTERNAL.ACTIVITY_TAG.ONE_ON_ONE) ?
+                SDK_EVENT.EXTERNAL.SPACE_TYPE.DIRECT :
+                SDK_EVENT.EXTERNAL.SPACE_TYPE.GROUP
+              // isRoomHidden,
+              // created,
+            };
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: This if block is really hard for me to grok. What is your line wrap length set on?

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 7]
```diff
@@ -415,7 +500,47 @@ const Memberships = SparkPlugin.extend({
...
+    const activity = {
+      id: deconstructHydraId(message.id).id
+    };
+    const conversation = {
+      id: deconstructHydraId(message.roomId).id
+    };
+
+    return this.spark.internal.conversation.acknowledge(conversation, activity)
+      .then((activity) => {
+        console.log(activity);
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: probably just for debugging

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js [hunk 8]
```diff
@@ -415,7 +500,47 @@ const Memberships = SparkPlugin.extend({
...
+          personEmail: activity.actor.emailAddress || activity.actor.entryEmail,
+          personDisplayName: activity.actor.displayName,
+          personOrgId: constructHydraId(hydraTypes.ORGANIZATION,
+            activity.actor.orgId),
+          roomId: constructHydraId(hydraTypes.ROOM, activity.target.id),
+          roomType:
+            activity.target.tags.includes(SDK_EVENT.INTERNAL.ACTIVITY_TAG.ONE_ON_ONE) ?
+              SDK_EVENT.EXTERNAL.SPACE_TYPE.DIRECT :
+              SDK_EVENT.EXTERNAL.SPACE_TYPE.GROUP,
+          isRoomHidden: false // any activity unhides a space.
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-memberships/src/memberships.js]: It might be a tad cleaner to have a function that you pass these objects too and based on keys, it properly encodes the things it needs to.

File: packages/node_modules/@webex/plugin-rooms/src/rooms.js [hunk 9]
```diff
@@ -241,6 +241,84 @@ const Rooms = SparkPlugin.extend({
...
+   * @memberof Rooms
+   * @returns {Promise<RoomObjectList>}
+   */
+  listWithReadStatus() {
+    return this.spark.internal.conversation.list({},
+      {
+        activitiesLimit: 0,
+        participantsLimit: 0,
+        isActive: true,
+        deferDecrypt: true
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-rooms/src/rooms.js]: with deferDecrypt don't you need to decrypt the payload somewhere?

File: .babelrc (hunk 1 above)
Comment (jpjpjp) [reply .babelrc]: This allows babel to support the "spread" operator.  It allows a syntax like this:
```
obj = {
  foo: source.fooey,
  ...(source.barey && {bar: source.barey})
};
```
Which results in an object that has no "bar" key if there is no "barey" in the source.
This syntax isn't NECESSARY, but the alternative is a bunch of if statements after the object definition, ie:

```
obj = {
 foo: source.fooey
};
if (source.barey) {
  obj.bar = source.barey;
}
```

Not terrible but if you have many keys that you want to evaluate the code gets much longer

File: packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js (hunk 2 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js]: Good point on the comment.  Will add it in the next commit.

The existing code always sets the the query param `includeParticipants=false`, which negates sending the participantAckFilter.  So if the caller explicitly sets the participantAckFilter we don't want that default setting.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 3 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Ack.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 5 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: This is exactly why I wanted you to review!  I'll discuss with @mattnorris if there is an established pattern for dealing with this.

File: .babelrc (hunk 1 above)
Comment (adamweeks) [reply .babelrc]: The widgets are using the ["transform-object-rest-spread"](https://www.npmjs.com/package/babel-plugin-transform-object-rest-spread) to accomplish this. Might make the change less intrusive to use that instead.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 6 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Basically I'm converting the conversations participant object into an external membership object.
linter throws a warning for lines longer than 80 so I added new lines in some places, does that ultimately make it harder to understand?

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 7 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Yes, thought I pulled those out before committing.   WIll update in next commit.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 8 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Do mean 
```
return ({
  key1 = makeKey1(activity),
  key2 = makeKey2(activity)
});
```
or
```
 return(convertActivityToMembershipSeen(activity));

File: packages/node_modules/@webex/plugin-rooms/src/rooms.js (hunk 9 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-rooms/src/rooms.js]: This is admitedly hacking on my part.   I "discovered" that if I set `deferDecrypt=true` that the response has the room title unencrypted in the `displayName` field for each conversation (that is not a one on one space).

File: packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js [hunk 10]
```diff
@@ -196,6 +198,89 @@ describe('plugin-rooms', function () {
       });
     });
 
+    describe('#listWithReadStatus()', () => {
+      // Reminder: we can't run the next two creates in parallel because some of
```
Comment (mattnorris) [inline packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js]: The tests should be independent so that we can isolate and skip tests to troubleshoot.

File: .babelrc (hunk 1 above)
Comment (jpjpjp) [reply .babelrc]: Ack, confirmed this works.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 5 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: As I look into this more deeply I see that the pattern of "assuming" hydra IDs is all over the "public" plugin module interfaces.   Will defer on solving this problem as part of this PR.

Comment (mattnorris) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Yes, for the public plugins we assume Hydra to reflect the Dev Portal.

File: packages/node_modules/@webex/plugin-memberships/test/integration/spec/memberships.js [hunk 11]
```diff
@@ -395,6 +451,63 @@ describe('plugin-memberships', function () {
...
     });
 
+    describe('#updateLastSeen()', () => {
+      let spark1, actor1, message, room;
+
+      before(() => {
+        // Get another SDK instance for the second user
+        // so that one user can send the message
+        // and another can mark it as read
+        spark1 = new CiscoSpark({credentials: user1.token});
```
Comment (mattnorris) [inline packages/node_modules/@webex/plugin-memberships/test/integration/spec/memberships.js]: If you need two different users and two different `spark`s, you can attach the `spark` to the user. 

```js
// where spock is the user
spock.spark = new CiscoSpark({...
```

File: .babelrc (hunk 1 above)
Comment (jpjpjp) [reply .babelrc]: @adamweeks this is in the latest commit now.

File: packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js (hunk 2 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/internal-plugin-conversation/src/conversation.js]: @andrewholsted  , comment added in latest commit.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 6 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Is this any better?

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 7 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: Gone now.

File: packages/node_modules/@webex/plugin-memberships/src/memberships.js (hunk 8 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/src/memberships.js]: I gave a cleanup technique a shot if you want to check it out again

File: .vscode/launch.json [hunk 12]
```diff
@@ -10,6 +10,21 @@
       "name": "Launch Chrome against localhost",
       "url": "https://localhost:8000",
       "webRoot": "${workspaceFolder}"
+    },
+    {
+      "type": "node",
+      "runtimeVersion": "8.15.1",
+      "request": "launch",
+      "name": "node test messages",
+      "runtimeExecutable": "/Users/jshipher/.nvm/versions/node/v10.10.0/bin/babel-node",
```
[the remark this answers is not in the document]
Comment (jpjpjp) [reply .vscode/launch.json]: Removed, checked in by accident.

File: package.json [hunk 13]
```diff
@@ -290,5 +290,8 @@
     "ignore": [
       "**/umd/**.js"
     ]
+  },
+  "devDependencies": {
+    "babel-preset-stage-0": "^6.24.1"
```
[the remark this answers is not in the document]
Comment (jpjpjp) [reply package.json]: Latest package.json should reflect Adams's suggestions.

File: packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js (hunk 10 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js]: Updated for independent tests.

File: packages/node_modules/@webex/plugin-memberships/test/integration/spec/memberships.js (hunk 11 above)
Comment (jpjpjp) [reply packages/node_modules/@webex/plugin-memberships/test/integration/spec/memberships.js]: Now implemented as you suggested

File: packages/node_modules/@webex/plugin-messages/src/messages.js [hunk 14]
```diff
@@ -241,29 +241,6 @@ const Messages = SparkPlugin.extend({
       .then((res) => res.body.items || res.body);
   },
 
-  /**
-   * Marks a message as read.
```
Comment (mattnorris) [inline packages/node_modules/@webex/plugin-messages/src/messages.js]: We don't want to be able to mark a message as read also? This is how the conversation API works.

File: packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js [hunk 15]
```diff
@@ -200,72 +206,77 @@ describe('plugin-rooms', function () {
...
-          spark2 = new CiscoSpark({credentials: user.token});
-
-          return spark2.people.get('me')
-            .then((person) => {
-              actor2 = person;
-              debug('SDK User (Actor) for tests:');
-              debug(`- name: ${actor.displayName}`);
-              debug(`-   id: ${actor2.id}`);
-            });
+          user2 = user;
```
Comment (mattnorris) [inline packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js]: It looks like you only need one user. If you don't want to use the cool name `spock` 😎 , do 

```
...
.then([u]) => {
  user = u;
...
```

Comment (mattnorris) [reply packages/node_modules/@webex/plugin-rooms/test/integration/spec/rooms.js]: nm, I see why you did this.

**Comment by mattnorris** [comment]
Failed to automatically close
